### PR TITLE
Suggest `split_at_mut` on multiple mutable index access

### LINKED ISSUE #58792: Suggest `split_at_mut` when trying to use multiple non-overlapping mutable slices to the same array
[The following code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1947bc965868f90b695a62803f999657) is invalid due to lifetime rules:

```
fn main() {
    let mut foo = [1,2,3,4];
    let mut a = &mut foo[..2];
    let mut b = &mut foo[2..];
    a[0] = 5;
    b[0] = 6;
    println!("{:?} {:?}", a, b);
}
```

```
error[E0499]: cannot borrow `foo` as mutable more than once at a time
 --> src/main.rs:4:22
  |
3 |     let mut a = &mut foo[..2];
  |                      --- first mutable borrow occurs here
4 |     let mut b = &mut foo[2..];
  |                      ^^^ second mutable borrow occurs here
5 |     a[0] = 5;
  |     ---- first borrow later used here
```

The solution is to either use `unsafe` or (more appropriately) use [`split_at_mut`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b789f2e5abc3895158f80a9602e59b6b):

```
fn main() {
    let mut foo = [1,2,3,4];
    let (mut a, mut b) = foo.split_at_mut(2);
    a[0] = 5;
    b[0] = 6;
    println!("{:?} {:?}", a, b);
}
```

The compiler should detect the case of multiple mutable borrows to the same array/slice and hint at the existence of `split_at_mut`. Ideally, we would also verify wether there's a range overlap in order to explain to the user why what they want to do is problematic and disallowed in Rust.

The output should ideally be along the lines of the following (the structured suggestion is not _needed_ to fix this issue, the text would be enough of an improvement):

```
error[E0499]: cannot borrow `foo` as mutable more than once at a time
 --> src/main.rs:4:22
  |
3 |     let mut a = &mut foo[..2];
  |                      --- first mutable borrow occurs here
4 |     let mut b = &mut foo[2..];
  |                      ^^^ second mutable borrow occurs here
5 |     a[0] = 5;
  |     ---- first borrow later used here
  = note: you cannot have multiple mutable borrows to the same piece of memory, as the compiler cannot assure that the different slices you're trying to use are non-overlapping
help: you can instead use `split_at_mut` to get two non-overlapping mutable slices from a single array/slice
  |
3 |     let (mut a, mut b) = foo.split_at_mut(2);
  |         ^^^^^^^^^^^^^^   ^^^^^^^^^^^^^^^^^^^
```

### PR DESCRIPTION
**Pull request by estebank**
cc #58792.

### CODE REVIEW

**Comment by estebank** [comment]
cc @oli-obk a (very) simplified version of the suggestion we discussed last RustConf.

File: src/test/ui/suggestions/suggest-split-at-mut.stderr [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+   |
+LL |     let a = &mut foo[2];
+   |             ----------- first mutable borrow occurs here
+LL |     let b = &mut foo[3];
+   |             ^^^^^^^^^^^ second mutable borrow occurs here
+LL |     *a = 5;
+   |     ------ first borrow later used here
+   |
+   = help: consider using `.split_at_mut(position)` or similar method to obtain two mutable non-overlapping sub-slices
+
```
Comment (memoryruins) [inline src/test/ui/suggestions/suggest-split-at-mut.stderr]: For short arrays, should we hint at patterns too? https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a9b151b57dc05010be68d1d714be741c

Comment (estebank) [reply src/test/ui/suggestions/suggest-split-at-mut.stderr]: I think that would be a good advice to give, but the compiler at this time doesn't know the possible size of the vector. I think that can be good follow up work.

**Comment by oli-obk** [comment]
Edited main post as to not close the issue yet. Still merging since this makes the `split_at_mut` methods so much more discoverable.

@bors r+

**Comment by HKalbasi** [comment]
I wonder why this PR works for `foo[2]` and `foo[3]` but doesn't work for `foo[..2]` and `foo[2..]` (so doesn't solve the #58792)? It seems compiler detects former as borrow of a projection (i.e. `foo[_]`) but the latter as borrow of `foo` itself. From my understanding they are two calls of Index trait so they shouldn't be different.

**Comment by estebank** [comment]
@HKalbasi Yes, I am not sure why there's divergence in the behavior. Maybe @oli-obk might know what's going on there.

**Comment by oli-obk** [comment]
`usize` indexing is a basic MIR operation, so the trait is not involved at all, while slicing requires going through the index trait, and thus you get a simple borrow and then a function call on that reference.

**Pull request merged**